virttest/migration.py

```diff
@@ -271,26 +283,187 @@ def do_migration(self, vms, srcuri, desturi, migration_type,
         :param thread_timeout: time out seconds for the migration thread running
         :param ignore_status: determine if an exception is raised for errors
         :param func: the function executed during migration thread is running
+        :param multi_funcs: list of functions executed during migration
+                     thread is running. The func and multi_funcs should not be
+                     provided at same time.
         :param args: dictionary used by func,
                      'func_param' is mandatory if no real func_param, none is
                      requested.
                      'shell' is optional, where shell=True(bool) can be used
                      for process.run
 
         """
+
+        def _run_collect_event_cmd():
+            """
+            To execute virsh event command to collect the domain events
+
+            :return: VirshSession to retrieve the events
+            """
+            cmd = "event --loop --all"
+            virsh_event_session = virsh.VirshSession(virsh_exec=virsh.VIRSH_EXEC,
+                                                     auto_close=True,
+                                                     uri=srcuri)
+            virsh_event_session.sendline(cmd)
+            logging.debug("Begin to collect domain events...")
+            return virsh_event_session
+
+        def _need_collect_events(funcs_to_run):
+            """
+            Check if there is a need to run a command to collect domain events
+
+            :param funcs_to_run: the functions to be run. It can be a list
+                                 or a single function
+                For example,
+                funcs_to_run = [{"func": <function check_established at 0x7f5833687510>,
+                                 "after_event": "iteration: '1'",
+                                 "before_event": "Suspended Migrated",
+                                 "before_pause": "yes"},
+                                {"func": <function domjobabort at 0x7f5835cd08c8>,
+                                 "before_pause": "no"}
+                               ]
+            :return: boolean, True to collect events, otherwise False
+            """
+            if not funcs_to_run:
+                return False
+            if isinstance(funcs_to_run, list):
+                for one_func in funcs_to_run:
+                    if isinstance(one_func, dict):
+                        after_event = one_func.get('after_event')
+                        before_event = one_func.get('before_event')
+                        if any([after_event, before_event]):
+                            return True
+                    else:
+                        # It should be a FunctionType
+                        return False
```

Comment:
As we discussed, let's raise an exception here. Thanks!

Reply:
Updated